lib/CL/pocl_mlir_build.cc

```diff
@@ -0,0 +1,296 @@
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#include "config.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/OwningOpRef.h"
+
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include <iostream>
+#include <llvm/Support/SourceMgr.h>
+
+#include "pocl/Transforms/Linker.h"
+
+#include <regex>
+#include <sstream>
+
+#include "pocl.h"
+#include "pocl_cache.h"
+#include "pocl_cl.h"
+#include "pocl_debug.h"
+#include "pocl_file_util.h"
+#include "pocl_llvm_api.h"
+#include "pocl_mlir.h"
+#include "pocl_mlir_file_util.hh"
+#include "pocl_util.h"
+
+static int getMLIRKernelLibrary(cl_device_id device,
+                                PoclLLVMContextData *llvm_ctx,
+                                mlir::OwningOpRef<mlir::ModuleOp> &lib) {
+  std::string kernellib_common, kernellib, kernellib_fallback;
+
+#ifdef ENABLE_POCL_BUILDING
+  if (pocl_get_bool_option("POCL_BUILDING", 0)) {
+    kernellib_common = BUILDDIR;
+    kernellib_common += "/lib/kernel/mlir";
+  } else // POCL_BUILDING == 0, use install dir
+#endif
+  {
+    char temp[POCL_MAX_PATHNAME_LENGTH];
+    pocl_get_private_datadir(temp);
+    kernellib_common = temp;
+  }
+
+  kernellib_common += "/";
+  kernellib = kernellib_common + "kernel-mlirbc";
+  kernellib += ".mlir";
+
+  if (device->kernellib_fallback_name) {
+    kernellib_fallback = kernellib_common + device->kernellib_fallback_name;
+    kernellib_fallback += ".mlir";
+  }
+
+  int read_status = 0;
+  if (pocl_exists(kernellib.c_str())) {
+    POCL_MSG_PRINT_LLVM("Using %s as the built-in lib.\n", kernellib.c_str());
+    read_status =
+        pocl_mlir_open_file(kernellib.c_str(), llvm_ctx->MLIRContext, lib);
+  } else {
+    if (device->kernellib_fallback_name &&
+        pocl_exists(kernellib_fallback.c_str())) {
+      POCL_MSG_WARN("Using fallback %s as the built-in lib.\n",
+                    kernellib_fallback.c_str());
+      read_status = pocl_mlir_open_file(kernellib_fallback.c_str(),
+                                        llvm_ctx->MLIRContext, lib);
+    } else
+      POCL_ABORT("Kernel library file %s doesn't exist.\n", kernellib.c_str());
+  }
+  return read_status;
+}
+
+static bool generateProgramMLIR(PoclLLVMContextData *Context,
+                                mlir::OwningOpRef<mlir::ModuleOp> &Mod,
+                                cl_device_id Device, std::string &Log) {
+
+  mlir::OwningOpRef<mlir::ModuleOp> BuiltinLib;
+  auto Error = getMLIRKernelLibrary(Device, Context, BuiltinLib);
+  if (Error) {
+    POCL_MSG_ERR("Failed retrieving kernel library with error code %d\n",
+                 Error);
+  }
+
+  if (MLIRLink(Mod, BuiltinLib, Log, Device->device_aux_functions,
+               Device->device_side_printf != CL_FALSE))
+    return true;
+
+  return false;
+}
+
+int pocl_mlir_build_program(cl_program program, unsigned device_i,
```

Comment:
`POCL_EXPORT` is missing here (MSVC requires it on both the declaration and the definition).